test/integration/apiserver/oidc/oidc_test.go

```diff
@@ -0,0 +1,505 @@
+			signingPrivateKey *rsa.PrivateKey,
+			caCertContent []byte,
+			caFilePath string,
+		)
+		configureOIDCServerBehaviour func(t *testing.T, oidcServer *utilsoidc.TestServer, signingPrivateKey *rsa.PrivateKey)
+		configureClient              func(
+			t *testing.T,
+			restCfg *rest.Config,
+			caCert []byte,
+			certPath,
+			oidcServerURL,
+			oidcServerTokenURL string,
+		) *kubernetes.Clientset
+		expectedErr error
+	}{
+		{
+			name:                    "ID token is ok",
+			configureInfrastructure: configureTestInfrastructure,
+			configureOIDCServerBehaviour: func(t *testing.T, oidcServer *utilsoidc.TestServer, signingPrivateKey *rsa.PrivateKey) {
+				idTokenLifetime := time.Second * 1200
+				oidcServer.TokenHandler().EXPECT().Token().Times(1).DoAndReturn(utilsoidc.TokenHandlerBehaviourReturningPredefinedJWT(
+					t,
+					signingPrivateKey,
+					oidcServer.URL(),
+					defaultOIDCClientID,
+					defaultOIDCClaimedUsername,
+					defaultStubAccessToken,
+					defaultStubRefreshToken,
+					time.Now().Add(idTokenLifetime).Unix(),
+				))
+			},
+			configureClient: configureClientFetchingOIDCCredentials,
+			expectedErr:     nil,
+		},
+		{
+			name:                    "ID token is expired",
+			configureInfrastructure: configureTestInfrastructure,
+			configureOIDCServerBehaviour: func(t *testing.T, oidcServer *utilsoidc.TestServer, signingPrivateKey *rsa.PrivateKey) {
+				configureOIDCServerToReturnExpiredIDToken(t, 2, oidcServer, signingPrivateKey)
+			},
+			configureClient: configureClientFetchingOIDCCredentials,
+			expectedErr:     new(apierrors.StatusError),
+		},
+		{
+			name:                    "client secret is not passed",
+			configureInfrastructure: configureTestInfrastructure,
+			configureOIDCServerBehaviour: func(t *testing.T, oidcServer *utilsoidc.TestServer, _ *rsa.PrivateKey) {
+				oidcServer.TokenHandler().EXPECT().Token().Times(2).Return(utilsoidc.Token{}, utilsoidc.ErrClientSecretNotPassed)
+			},
+			configureClient: configureClientWithEmptyIDToken,
+			expectedErr:     utilsoidc.ErrClientSecretNotPassed,
+		},
+		{
+			name:                    "client secret is wrong",
+			configureInfrastructure: configureTestInfrastructure,
+			configureOIDCServerBehaviour: func(t *testing.T, oidcServer *utilsoidc.TestServer, _ *rsa.PrivateKey) {
+				oidcServer.TokenHandler().EXPECT().Token().Times(2).Return(utilsoidc.Token{}, utilsoidc.ErrBadClientSecret)
+			},
+			configureClient: configureClientWithEmptyIDToken,
+			expectedErr:     utilsoidc.ErrBadClientSecret,
+		},
+		{
+			name:                    "wrong client ID",
+			configureInfrastructure: configureTestInfrastructure,
+			configureOIDCServerBehaviour: func(t *testing.T, oidcServer *utilsoidc.TestServer, _ *rsa.PrivateKey) {
+				oidcServer.TokenHandler().EXPECT().Token().Times(2).Return(utilsoidc.Token{}, utilsoidc.ErrBadClientID)
+			},
+			configureClient: configureClientWithEmptyIDToken,
+			expectedErr:     utilsoidc.ErrBadClientID,
+		},
```

Comment:
These 3 tests are essentially the same with the only difference being the error returned by the OIDC server? IIUC, the only thing we're really validating here is the error from the OIDC server is funneled all the way down?

Reply:
The idea is to check that API Server returns the same error which it gets from the OIDC Server. But agreed, that one test is enough, I left only `"wrong client ID"`

Reply:
> If that's the case, could we use [t.TempDir()](https://pkg.go.dev/testing#T.TempDir) that'll automatically get cleaned up after tests and subtests are complete?

Yeah, we already use it, but I got used to cleanup any artifacts explicitly and don't rely on magic under OS or etc., do you consider that's redundant?